### PR TITLE
RNN 循环神经网络系列 2：文本分类

### LINKED ISSUE #2248: RNN 循环神经网络系列 2：文本分类
* 原文链接：[RECURRENT NEURAL NETWORKS (RNN) – PART 2: TEXT CLASSIFICATION](https://theneuralperspective.com/2016/10/06/recurrent-neural-networks-rnn-part-2-text-classification/)
  * Markdown文件：[文件地址](https://github.com/xitu/gold-miner/blob/master/TODO/recurrent-neural-networks-rnn-part-2-text-classification.md)
  * PR 地址：#2270
  * 文章分类：**机器学习**
  ----
  * 翻译时间：3 天
  * 校对时间：1 天
  * 翻译积分：3 分
  * 校对积分：1 分
  * [积分有什么用](https://github.com/xitu/gold-miner/wiki)

### PR DESCRIPTION
**Pull request by changkun**
译文翻译完成，resolve #2248

### CODE REVIEW

**Comment by yanqiangmiffy** [comment]
校对认领

**Comment by fanyijihua** [comment]
@yanqiangmiffy 好的呢 :beer:

File: TODO/recurrent-neural-networks-rnn-part-2-text-classification.md [hunk 1]
```diff
@@ -138,70 +138,69 @@ class model(object):
                 return outputs[0], outputs[1]
             elif sampling:
                 return outputs[0]
-        else: # training
+        else: # 训练
             return outputs[0], outputs[1], outputs[2]
 ```
 
-The code above is the model that trains using the input text. **Note**: We decided to keep batch size in our input and target placeholders for clarity but we should make these independent of a particular batch size. If we do this, we will have to feed in initial_state since it depends on batch_size. We feed in the tokens from each input sequence by embedding them. A proven strategy for better performance is to pretrain these embedding weights with a skip-gram model on the input text.
+上面的代码就是我们的模型代码，它在训练的过程中使用了输入的文本。**注意**：为了清楚期间，我们决定将批量数据的大小保存在我们的输入和目标占位符中，但是我们应该让它们独立于一个特定的批量大小之外。由于这个特定的批量大小依赖于 `batch_size`，如果我们这么做，那么我们就还得输入一个 `initial_state`。我们通过嵌入他们来为每个数据序列来输入 token。实践策略表明，在输入文本上预训练这些包含连续跳跃元模型的嵌入权重能够获得更好的性能。
```
Comment (yanqiangmiffy) [inline TODO/recurrent-neural-networks-rnn-part-2-text-classification.md]: 『在输入文本上预训练这些包含连续跳跃元模型的嵌入权重能够获得更好的性能』=>『我们在输入文本上使用使用skip-gram模型预训练嵌入权重能够取得更好的性能』

File: TODO/recurrent-neural-networks-rnn-part-2-text-classification.md [hunk 2]
```diff
@@ -138,70 +138,69 @@ class model(object):
...
 
-In this model, we use dynamic_rnn again but this time we feed in the value for sequence_length which is a list that has the length of each sequence. By doing this, we avoid doing unnecessary computation past the last word of the input sequence. The function **length** is what gives us this list and is shown below. We could’ve also just as easily calculated seq_len outside and then pass it in via a placeholder.
+在此模型中，我们再次使用 `dynamic_rnn`，但是这次我们提供了`sequence_length` 参数的值，它是一个包含每个序列长度的列表。这样，我们就可以避免在输入序列的最后一个字之后不必要的计算。**`length`** 函数就用来计算这个列表，如下所示。当然，我们也可以在外面计算`seq_len`，再通过占位符进行传递。
```
Comment (yanqiangmiffy) [inline TODO/recurrent-neural-networks-rnn-part-2-text-classification.md]: 『**`length`** 函数就用来计算这个列表』=>『**`length`** 函数就用来获取这个列表的长度』

File: TODO/recurrent-neural-networks-rnn-part-2-text-classification.md [hunk 3]
```diff
@@ -138,70 +138,69 @@ class model(object):
...
+```python
 def length(data):
 	relevant = tf.sign(tf.abs(data))
 	length = tf.reduce_sum(relevant, reduction_indices=1)
 	length = tf.cast(length, tf.int32)
 	return length
 ```
 
-Since our PAD token is a 0, we can use the sign property of each token to determine wether it is a padding token or not. tf.sign will give us a 1 if input > 0 and 0 if input==0\. By using this, we can find the number of tokens with the positive sign by reduce by the column index. We can now feed this length into dynamic_rnn.
+由于我们填充符 token 为 0，因此可以使用每个 token 的 sign 性质来确定它是否是一个填充符 token。如果输入大于 0，则 `tf.sign` 为 1；如果输入为 0，则为 `tf.sign` 为 0。这样，我们可以通过 sign 值为正这个性质来通过列索引逐个获得 token 的数量。至此，我们可以将这个长度提供给 `dynamic_rnn` 了。
```
Comment (yanqiangmiffy) [inline TODO/recurrent-neural-networks-rnn-part-2-text-classification.md]: 『我们可以通过 sign 值为正这个性质来通过列索引逐个获得 token 的数量』=>『我们可以逐步通过列索引来获得sign值为正的token数量』

**Comment by leviding** [comment]
校对姿势完美哈哈 @yanqiangmiffy

**Comment by TobiasLee** [comment]
@fanyijihua  校对认领

**Comment by fanyijihua** [comment]
@TobiasLee 妥妥哒 :beers:

File: TODO/recurrent-neural-networks-rnn-part-2-text-classification.md [hunk 4]
```diff
@@ -3,111 +3,111 @@
...
-2. Convert sentences into numeric tokens.
-3. Store sequence lengths for each sentence
+1. 清洗句子并切分成一个个 token；
+2. 将句子转换为数值 token；
+3. 保存每个句子的序列长。
 
 ![Screen Shot 2016-10-05 at 7.32.36 PM.png](https://theneuralperspective.files.wordpress.com/2016/10/screen-shot-2016-10-05-at-7-32-36-pm.png?w=620)
 
-As you can see in the diagram above, we want to predict the sentiment of the sentence as soon as it is complete. Factoring in the additional paddings will introduce too much noise and your network will not perform well. **Note**: The only reason we pad our sequences is because we need to feed in fixed sized batches into the RNN. As you will see below, using a dynamic rnn will allow us to prevent unnecessary calculations once the sequence is complete.
+如上图所示，我们希望在计算完成时立即对句子的情绪做出预测。引入额外的填充符会带来过多噪声，这样的话你网络的性能就会不太好。**注意**：我们填充序列的唯一原因是因为需要以固定大小的批量输入进 RNN。下面你会看到，使用动态 RNN 还能避免在序列完成后的不必要计算。
```
Comment (TobiasLee) [inline TODO/recurrent-neural-networks-rnn-part-2-text-classification.md]: 引入额外的填充符会带来过多噪声，这样的话你网络的性能就会不太好 =>
引入额外的填充符会带来过多噪声，这样的话你模型的性能就会不太好
”网络“改成”模型“可以避免一些歧义

File: TODO/recurrent-neural-networks-rnn-part-2-text-classification.md [hunk 5]
```diff
@@ -3,111 +3,111 @@
...
 
-        # Process RNN outputs
+        # 处理 RNN 输出
         with tf.variable_scope('rnn_softmax'):
             W_softmax = tf.get_variable("W_softmax",
                 [FLAGS.num_hidden_units, FLAGS.num_classes])
             b_softmax = tf.get_variable("b_softmax", [FLAGS.num_classes])
 
-        # Logits
+        # Logit
```
Comment (TobiasLee) [inline TODO/recurrent-neural-networks-rnn-part-2-text-classification.md]: Logit =>
Logits

File: TODO/recurrent-neural-networks-rnn-part-2-text-classification.md [hunk 6]
```diff
@@ -3,111 +3,111 @@
...
         logits = rnn_softmax(FLAGS, outputs)
         probabilities = tf.nn.softmax(logits)
         self.accuracy = tf.equal(tf.argmax(
             self.targets_y,1), tf.argmax(logits,1))
 
-        # Loss
+        # 损失函数
         self.loss = tf.reduce_mean(
             tf.nn.sigmoid_cross_entropy_with_logits(logits, self.targets_y))
```
Comment (TobiasLee) [inline TODO/recurrent-neural-networks-rnn-part-2-text-classification.md]: self.loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(logits, self.targets_y)) =>
self.loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(logits=logits, labels=self.targets_y))
新版 Tensorflow 的 cross entropy 方法强制需要传递关键字参数，为了避免读者实际运行时出现 bug，添加上可能更为妥当

File: TODO/recurrent-neural-networks-rnn-part-2-text-classification.md [hunk 7]
```diff
@@ -3,111 +3,111 @@
...
+        # (在每个单词后生成情绪)
 
-        # this is taking all the ouputs for the first input sequence
-        # (only 1 input sequence since we are sampling)
+        # 取所有输出作为第一个输入序列
+        # (由于采样，只需一个输入序列)
         sampling_outputs = all_outputs[0]
 
-        # Logits
+        # Logit
```
Comment (TobiasLee) [inline TODO/recurrent-neural-networks-rnn-part-2-text-classification.md]: Logit =>
Logits

File: TODO/recurrent-neural-networks-rnn-part-2-text-classification.md [hunk 8]
```diff
@@ -138,70 +138,69 @@ class model(object):
                 return outputs[0], outputs[1]
             elif sampling:
                 return outputs[0]
-        else: # training
+        else: # 训练
             return outputs[0], outputs[1], outputs[2]
 ```
 
-The code above is the model that trains using the input text. **Note**: We decided to keep batch size in our input and target placeholders for clarity but we should make these independent of a particular batch size. If we do this, we will have to feed in initial_state since it depends on batch_size. We feed in the tokens from each input sequence by embedding them. A proven strategy for better performance is to pretrain these embedding weights with a skip-gram model on the input text.
+上面的代码就是我们的模型代码，它在训练的过程中使用了输入的文本。**注意**：为了清楚期间，我们决定将批量数据的大小保存在我们的输入和目标占位符中，但是我们应该让它们独立于一个特定的批量大小之外。由于这个特定的批量大小依赖于 `batch_size`，如果我们这么做，那么我们就还得输入一个 `initial_state`。我们通过嵌入他们来为每个数据序列来输入 token。实践策略表明，我们在输入文本上使用 skip-gram 模型预训练嵌入权重能够取得更好的性能。
```
Comment (TobiasLee) [inline TODO/recurrent-neural-networks-rnn-part-2-text-classification.md]: 为了清楚期间 =>
为了清楚起见

File: TODO/recurrent-neural-networks-rnn-part-2-text-classification.md [hunk 9]
```diff
@@ -138,70 +138,69 @@ class model(object):
...
 
-In this model, we use dynamic_rnn again but this time we feed in the value for sequence_length which is a list that has the length of each sequence. By doing this, we avoid doing unnecessary computation past the last word of the input sequence. The function **length** is what gives us this list and is shown below. We could’ve also just as easily calculated seq_len outside and then pass it in via a placeholder.
+在此模型中，我们再次使用 `dynamic_rnn`，但是这次我们提供了`sequence_length` 参数的值，它是一个包含每个序列长度的列表。这样，我们就可以避免在输入序列的最后一个字之后不必要的计算。**`length`** 函数就用来获取这个列表的长度，如下所示。当然，我们也可以在外面计算`seq_len`，再通过占位符进行传递。
```
Comment (TobiasLee) [inline TODO/recurrent-neural-networks-rnn-part-2-text-classification.md]: 这样，我们就可以避免在输入序列的最后一个字之后不必要的计算 =>
这样，我们就可以避免在输入序列的最后一个词之后进行的不必要的计算
”词“一般是序列的最小单位，增加”进行的“读起来可能更为通顺一些

File: TODO/recurrent-neural-networks-rnn-part-2-text-classification.md [hunk 10]
```diff
@@ -138,70 +138,69 @@ class model(object):
...
+**注意**：这是一个非常简单的模型，其数据集非常有限。主要目的只是为了阐明它是如何搭建以及如何运行的。为了获得更好的性能，请尝试使用数据量更大的数据集，并考虑具体的网络架构，比如 Attention 模型、Concept-Aware 词嵌入以及隐喻（symbolization to name）等等。
 
-## Extra: Loss Masking (not needed here)
+## 损失屏蔽（这里不需要）
 
-Finally, we compute the cost and you may notice we aren’t doing any “loss masking” here because we isolated our relevant output and used only that to compute the loss. However, for other tasks such as machine translation, we may have the outputs that come from the PAD tokens as well. We do not want to account for these outputs because the dynamic_rnn with seq_lens parameter fulfilled will return 0 for these outputs. Here is a simple example of what this implementation will look like; again, we use the fact that our PAD token is a 0.
+最后，我们来计算 cost。你可能会注意到我们没有做任何损失屏蔽（loss masking）处理，因为我们分离了对应输出，仅用于计算损失函数。然而，对于其他诸如机器翻译的任务来说，我们的输出很有可能还来自填充符 token。我们不想考虑这些输出，因为传递了 `seq_lens` 参数的 `dynamic_rnn` 将返回 0。下面这个例子比较简单，只用来说明这个实现大概是怎么回事；再次，我们使用填充符 token 为 0 的性质：
```
Comment (TobiasLee) [inline TODO/recurrent-neural-networks-rnn-part-2-text-classification.md]: 再次，我们使用填充符 token 为 0 的性质： =>
我们这里再一次使用了填充符 token 为 0 的性质：
可能更符合中文的语序

File: TODO/recurrent-neural-networks-rnn-part-2-text-classification.md [hunk 11]
```diff
@@ -138,70 +138,69 @@ class model(object):
...
 
-```
-# Flatten the logits and targets
-targets = tf.reshape(targets, [-1]) # flatten targets
+```python
+# 向量化 logit 和目标
```
Comment (TobiasLee) [inline TODO/recurrent-neural-networks-rnn-part-2-text-classification.md]: 向量化 logit 和目标 =>
向量化 logits 和目标标签

File: TODO/recurrent-neural-networks-rnn-part-2-text-classification.md [hunk 12]
```diff
@@ -138,70 +138,69 @@ class model(object):
...
 
-Now we are going to feed into softmax weights but before that we convert state from **[1, N, D]** to **[N, D]** by just taking the first index (state[0]). We can now dot product with softmax weights **[D, C]** to get outputs of **[N, C]** which we do the exponential softmax operation and normalization to and compute the loss with the target_y **[N, C]**.
+我们需要将这个嵌入后的 `X` 传递给 `dynamic_rnn` 并返回 `all_outputs` （`[N, <max_len>, D]`）以及 `state`（`[1, N, D]`）。由于我们输入了 `seq_lens`，对于我们而言它就是最后一个对应的状态。从维度的角度来说，你可以看到， `all_outputs` 就是来自 RNN 的对于每个句子中的每个词的全部输出结果。然而，`state` 仅仅只是每个句子的最后一个对应输出。
 
-**Note**: If you used an basic RNN or GRU, the shapes of the all_outputs and state returns from dynamic_rnn would be same. But if we used an LSTM the all_outputs still has shape **[N, <max_len>, D]** but state is of shape **[1, 2, N, D]**.
+现在我们要输入 softmax 权重，但在此之前，我们需要通过取第一个索引（`state[0]`）来把状态从 `[1,N,D]` 转换为`[N,D]`。如此便可以通过与 softmax 权重 `[D,C]` 的点积，来得到形状为 `[N,C]` 的输出。其中，我们做指数级 softmax 运算，然后进行正规化，最终结合形状为 `[N,C]` 的 `target_y` 来计算损失函数。
```
Comment (TobiasLee) [inline TODO/recurrent-neural-networks-rnn-part-2-text-classification.md]: 然后进行正规化 =>
然后进行正则化

**Comment by changkun** [comment]
@TobiasLee 非常感谢相关校对建议和意见！关于 `tf.nn.sigmoid_cross_entropy_with_logits ` 的显式参数名我没有做更改。有几个原因，其实这五篇文章的代码我都有下载下来运行，这系列博文的作者在写相关代码的时候是基于 py2 和 tf 0.12 之前的版本，代码中有相当多的「过时」函数，这些要么是被弃用了、要么是参数顺序发生了变化、要么是函数名做了修改，`tf.nn.sigmoid_cross_entropy_with_logits ` 是其中之一，这其中还有比如说 `tf.concat` `tf.nn.rnn_cell._linear` `tf.pack`等等，想要跑通原代码还需要做除了cross entropy 之外的很多适配修改，会引入更多的工作量..🤣 考虑到原文所写相关代码的只是为了方便读者理解相关概念，所以干脆统一就不改了吧?

**Comment by leviding** [comment]
已经 merge 啦~ 快快麻溜发布到掘金专栏然后给我发下链接，方便及时添加积分哟。

掘金翻译计划有自己的知乎专栏，你也可以投稿哈，推荐使用一个[好用的插件](https://chrome.google.com/webstore/detail/minimalist-markdown-edito/pghodfjepegmciihfhdipmimghiakcjf?hl=zh-CN)。
专栏地址：https://zhuanlan.zhihu.com/juejinfanyi

**Comment by leviding** [comment]
@changkun 发布文章反馈链接啦

**Comment by changkun** [comment]
@leviding sorry... 发完文章后就搞忘了... https://juejin.im/post/59f0c6b3f265da4319557de4
